src/modules/csm/csm.py

```diff
@@ -0,0 +1,308 @@
+                except KeyError:
+                    # It's possible that the validator is not assigned to any duty, hence it's performance
+                    # is not presented in the aggregates (e.g. exited, pending for activation etc).
+                    continue
+
+                if aggr.perf > threshold:
+                    # Count of assigned attestations used as a metrics of time
+                    # the validator was active in the current frame.
+                    distribution[no_id] += aggr.assigned
+
+        # Calculate share of each CSM node operator.
+        shares = defaultdict[NodeOperatorId, int](int)
+        total = sum(p for p in distribution.values())
+
+        if not total:
+            return 0, shares
+
+        to_distribute = self.w3.csm.fee_distributor.shares_to_distribute(blockstamp.block_hash)
+        for no_id, no_share in distribution.items():
+            if no_share:
+                shares[no_id] = to_distribute * no_share // total
+
+        distributed = sum(s for s in shares.values())
+        assert distributed <= to_distribute
+        return distributed, shares
+
+    def stuck_operators(self, blockstamp) -> Iterable[NodeOperatorId]:
+        l_ref_slot, _ = self.current_frame_range(blockstamp)
+        # NOTE: r_block is guaranteed to be <= ref_slot, and the check
+        # in the inner frames assures the  l_block <= r_block.
+        return self.w3.csm.get_csm_stuck_node_operators(
+            get_first_non_missed_slot(
+                self.w3.cc,
+                l_ref_slot,
+                blockstamp.slot_number,
+                direction='forward',
+            ).message.body.execution_payload.block_hash,
+            blockstamp.block_hash,
+        )
+
+    @lru_cache(maxsize=1)
+    def current_frame_range(self, blockstamp: BlockStamp) -> tuple[SlotNumber, SlotNumber]:
+        converter = self.converter(blockstamp)
+
+        far_future_initial_epoch = converter.get_epoch_by_timestamp(FAR_FUTURE_EPOCH)
+        if converter.frame_config.initial_epoch == far_future_initial_epoch:
+            raise ValueError("CSM oracle initial epoch is FAR_FUTURE_EPOCH")
+
+        l_ref_slot = self.w3.csm.get_csm_last_processing_ref_slot(blockstamp)
+        r_ref_slot = self.get_current_frame(blockstamp).ref_slot
+
+        # The very first report, no previous ref slot.
+        if not l_ref_slot:
+            l_ref_slot = SlotNumber(self.get_initial_ref_slot(blockstamp) - converter.slots_per_frame)
+
+        # We are between reports, next report slot didn't happen yet. Predicting the next ref slot for the report
+        # to calculate epochs range to collect the data.
+        if l_ref_slot == r_ref_slot:
+            r_ref_slot = converter.get_epoch_last_slot(
+                EpochNumber(converter.get_epoch_by_slot(l_ref_slot) + converter.frame_config.epochs_per_frame)
+            )
+
+        return l_ref_slot, r_ref_slot
```

Comment:
l_ref_slot should be included in report
add +1

Reply:
we already do that in `collect_data` when calculating `l_epoch`